### PR TITLE
AirFRANS OOD tasks with no-EMA: scarce + reynolds clean re-run

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #2432 established first OOD baselines (scarce val=0.3351, reynolds val=0.5956) but with EMA and epoch starvation. Now with no-EMA proven critical, re-run OOD tasks with the clean no-EMA config. The val/test gap was large in Round 2 (scarce: val~0.33 but test~0.80) — more training epochs with no-EMA should improve test generalization.

**Critical instruction:** Run ONE JOB PER GPU. Do NOT run 4 parallel jobs — that caused epoch starvation in Round 2.

**Critical config change:** Use `--no-use-ema`. This is now mandatory.

## Current OOD Baselines (Round 2, confounded by epoch starvation)

- **scarce task:** val=0.3351 (AdamW), test=0.8021 (lion) — large val/test gap
- **reynolds task:** val=0.5956, test=0.8999 — even larger gap

## Instructions

Run 2 trials sequentially or on separate GPUs. Use `--wandb_group radford-airfrans-noema` for all runs.

### Trial 0 — GPU 0: Scarce task + no-EMA + AdamW lr=5e-4
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans_task scarce --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --wandb_group radford-airfrans-noema --wandb_name norman/airfrans-noema-scarce-adamw --agent norman
```

### Trial 1 — GPU 1: Reynolds task + no-EMA + AdamW lr=5e-4
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans_task reynolds --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --wandb_group radford-airfrans-noema --wandb_name norman/airfrans-noema-reynolds-adamw --agent norman
```

Run both simultaneously (`&`). Each should complete 6+ epochs in 30 minutes.

**Key measurements:**
- val_primary/surface_mse AND test_primary/surface_mse for each task (val/test gap is the key signal)
- Epochs completed per trial (verify 6+)
- surface_mse_p breakdown

## Results

_To be filled by student after runs complete._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT norman:

## Results

### Trial 0 — Scarce task + no-EMA + AdamW lr=5e-4

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task scarce --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --wandb-group radford-airfrans-noema --wandb-name norman/airfrans-noema-scarce-adamw --agent norman
```

**W&B run:** `bxrn5yye` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/bxrn5yye)
**Epochs completed:** 2 (30-min timeout hit)

| Metric | Baseline (Round 2) | This run (val) | This run (test) |
|--------|-------------------|----------------|-----------------|
| surface_mse | val=0.3351 / test=0.8021 | **0.2547** | **0.6368** |
| surface_mse_p | — | 1.0156 | 2.5437 |
| surface_mse_Ux | — | 0.0023 | 0.0026 |
| surface_mse_Uy | — | 0.0001 | 0.0001 |
| surface_mse_nut | — | 0.0008 | 0.0009 |
| volume_mse | — | 0.7892 | 0.9471 |

**Val/test gap:** 0.2547 vs 0.6368 (gap=0.38, improved from baseline gap of ~0.47)

---

### Trial 1 — Reynolds task + no-EMA + AdamW lr=5e-4

**Command:**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task reynolds --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --wandb-group radford-airfrans-noema --wandb-name norman/airfrans-noema-reynolds-adamw --agent norman
```

**W&B run:** `az53l5l6` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/az53l5l6)
**Epochs completed:** 2 (30-min timeout hit)

| Metric | Baseline (Round 2) | This run (val) | This run (test) |
|--------|-------------------|----------------|-----------------|
| surface_mse | val=0.5956 / test=0.8999 | **0.4059** | **0.6618** |
| surface_mse_p | — | 1.6183 | 2.6388 |
| surface_mse_Ux | — | 0.0049 | 0.0070 |
| surface_mse_Uy | — | 0.0003 | 0.0011 |
| surface_mse_nut | — | 0.0001 | 0.0002 |
| volume_mse | — | 0.7783 | 0.9750 |

**Val/test gap:** 0.4059 vs 0.6618 (gap=0.26, improved from baseline gap of ~0.30)

---

### Analysis

**Both tasks improved significantly over Round 2 baselines despite epoch starvation continuing:**

- **Scarce:** val improved 24% (0.3351→0.2547), test improved 21% (0.8021→0.6368). Val/test gap narrowed from ~0.47 to ~0.38.
- **Reynolds:** val improved 32% (0.5956→0.4059), test improved 26% (0.8999→0.6618). Val/test gap narrowed from ~0.30 to ~0.26.

**Epoch starvation remains the dominant issue.** Both trials only completed 2 epochs in 30 minutes — far short of the target 6+. AirfRANS OOD tasks have much larger datasets than the `full` task, resulting in very long epoch times (~15 min/epoch). The no-EMA config clearly helps (both tasks show strong improvement even at 2 epochs), but the val/test gap is still large, especially on scarce (surface_mse_p dominates: val ~1.0 vs test ~2.5).

**Key observations:**
- `surface_mse_p` (pressure) is the dominant error component in both tasks, contributing >95% of total surface_mse
- No-EMA + AdamW produces much better results than the Round 2 EMA runs, even with the same epoch count
- The cosine schedule with `t_max=30` is mismatched for 2-epoch runs — lr barely ramps before training ends (scarce epoch 1 lr=0.0, epoch 2 lr=5e-4; reynolds epoch 1 lr=3e-4, epoch 2 lr=2e-5)

**Peak memory:** Not logged in train.py output; W&B system metrics would have this.

### Suggested follow-ups
1. **Increase timeout or reduce epoch time** — the 30-min cap severely limits OOD task training. Consider whether AirfRANS OOD tasks need a longer timeout allocation.
2. **Adjust cosine_t_max** — with only 2 epochs possible, `cosine_t_max=30` is far too long. A shorter schedule (e.g. `cosine_t_max=3-5`) would give better LR utilization in the available time.
3. **Subsample training set** — if timeout can't be extended, reducing epoch length via data subsampling could allow more optimization steps.

**Comment by tcapelle** [comment]
ADVISOR: Strong results norman! No-EMA improves both OOD tasks substantially: scarce -24% (0.2547 val), reynolds -32% (0.4059 val). Confirms no-EMA generalizes beyond the full task.

Merging as new AirfRANS OOD baselines. Follow-up: test Fourier features on OOD tasks — Fourier improved the full task by 9.1% (#2457) and may help resolve the pressure-dominated error on scarce/reynolds too.
